### DISCUSSION TITLE
Allow static file Cache-Control TTL to be specified

### DISCUSSION THREAD

**Discussion by stevethatcodes** [vercel/next.js#40896] [Ideas]
### Describe the feature you'd like to request

The bundled JavaScript files contained within `.next/static` are currently served up with the following `Cache-Control` header:

```
Cache-Control: public, max-age=31557600, immutable
```

Whilst the way in which these bundles are generated, ensures that unless their content changes, the filename does not, meaning they _can_ be cached effectively forever, this can present challenges when used at scale in production. Considering the following scenario:

If one of the JavaScript bundles ends up being served badly for any reason, for example is truncated or malformed, and that file ends up in cache, Browser or Edge network, it causes the NextJS app to break in an often hard-to-observe way. Worst case, if it's in the Browser cache like that, you either have to change the bundle and redeploy the NextJS app, or, ask the end-user to clear their cache, which is not always an option especially when in production and at scale.

By allowing the TTL to be configurable, or even the entire Cache-Control header for these files, it gives important control over the caching behaviour of those files.

### Describe the solution you'd like

Within `next.config.js` there could be a property that allows the configuration of the TTL for the static bundles, e.g. 

```
staticBundleTTL: 3600 // Default 31557600
```

Alternatively the value of the header could be provided in full:

```
staticBundleCacheControl: 'public, max-age=7200, must-revalidate' // Default to public, max-age=31557600, immutable
```

### Describe alternatives you've considered

Override the header in another layer of infrastructure, e.g. Reverse Proxy or Edge network.

Whilst this is a viable option, it's added complexity and introduces a second source of truth to response headers coming from NextJS.

Additionally, some solutions for example Cloudflare's Browser TTL override, only allow you to specify a value greater than the existing max-age.

**Comment by SukkaW** [comment]
It is not necessary. Once you deploy a new build, the old bundles from the previous build **will never ever be served again**.

This is because the filenames of emitted bundles depend on the content hash. If your source code has changed, it is guaranteed that the filename will change.

Not only that, every time Next.js builds, it will generate a unique id (for each build) only. The build id is used when serving JSONs (from `getStaticProps` only. JSON from `getServerSideProps` will be served under `_next/data` endpoint, not `_next/static`). This is to make sure for every build, the URL of served `getStaticProps` will always change (even if the return value of `getStaticProps` doesn't change at all!).

So it is impossible for Next.js to serve stale files. You have corrupted files? You only need to fix it and deploy a new build, and Next.js will make sure previous files will never be served again.

**Comment by stevethatcodes** [comment]
All that makes sense, the problem isn't getting Next.js to serve up fresh files though. If there's a corrupt file that's ended up cached somewhere, either at the Edge or in the Browser, it won't attempt to request it from Next.js again. 

If I have a bundle that seldom changes, but becomes corrupt, currently I either have to force a change to that bundle to get it a fresh filename, or have affected users clear their cache.

The files in question are those loaded by the main HTML, via the `_next/static/chunks` path.

Granted the existing behaviour supports the concept that these bundles _should_ never change and are safe to cache indefinitely, this feature request is to allow that to be configured for instances where a lower TTL is more appropriate.

**Comment by SukkaW** [comment]
> All that makes sense, the problem isn't getting Next.js to serve up fresh files though. If there's a corrupt file that's ended up cached somewhere, either at the Edge or in the Browser, it won't attempt to request it from Next.js again.

If you have changed the source code somewhere, the hash of the bundle changes, so the file name of the bundle changes, which also means the filename of the webpack manifest will also changes. Since the HTML is never cached, the browser will guaranteed to use the newest webpack manifest to fetch the newest available bundle.

As long as the source code changes, the previous bundle will never ever be served again.
 
> If I have a bundle that seldom changes, but becomes corrupt, currently I either have to force a change to that bundle to get it a fresh filename, or have affected users clear their cache.

What's the reason behind the corruption?

If your site is deployed on Vercel, **you should contact Vercel support immediately**, as it indicates something went really wrong in Vercel's infrastructure.

If your site is deployed on your own server and you are using a third-party CDN, it is possible that your CDN failed to fetch the bundle from your origin server. Although it will almost never happen, you only have to purge the CDN cache in case this does happen.

To prevent your CDN from failing to connect to your origin server, you could also:

- Deploy the `_next/static` folder to the object storage, and use it as the origin server of your CDN.
- Enable the origin shield / tiered cache feature for your CDN. It will reduce the requests between the CDN and your origin server, thus reducing the chance of corruption.

Since you are already using a CDN, it would be very easy to override the `cache-control` anyway.

**Comment by stevethatcodes** [comment]
> If you have changed the source code somewhere, the hash of the bundle changes, so the file name of the bundle changes, which also means the filename of the webpack manifest will also changes. Since the HTML is never cached, the browser will guaranteed to use the newest webpack manifest to fetch the newest available bundle.

This requires the content to change as you say, it's not always practical to cause a bundle's content to change just to give it a new filename.

> What's the reason behind the corruption?

The reason for corruption / truncation isn't really important here. In my case it was part of a network stack returning a partial file. The point is that once a file's present in caches, it's not always as easy as purging caches to fix the issue.

> you only have to purge the CDN cache in case this does happen

This does not solve the problem of a file being cached in a browser. Nor does the concept of simply changing code in a bundle, as mentioned above that's not always practical to do.

> Since you are already using a CDN, it would be very easy to override the cache-control anyway.

This is not necessarily true and I'd argue that being able to control the header at source, not on the edge should always be the preferred approach. If Edge network or CDN or infrastructure is changed, it's another part that needs to be maintained outside of Next.js.

All of your commentary is entirely fair, however my perspective on this is that having this be configurable is inline with _many_ of the parts of Next.js that are already configurable, for example gzip, keep-alive, headers (outside of Cache-Control).

**Reply by romiohasan** [reply]
I am in a situation now that needs the proposed solution just as @stevethatcodes  mentioned.

I have already spend last two workdays to solve this sudden issue where no matter what solution I tried, users who have already visited our production site, always gets an old version of the content.
We have thousands of users and we can not simply ask them to clear their cache. 
Our web app is deployed with AWS with no CDN configured.

Here is a simple overview of the problem :
App is deployed with new features on live production server and staging server.
(1) Users who have already visited live production server are not getting the new features 
( unless they open the site in the incognito mode or visit from different browser without old cached version).
(2) visiting staging server always loads the new feature regardless of the state of the user browser cache.

Already redeployed production server several times but the problem remains the same for users with old browser cache.
I have added the screenshot below that shows the state of those files :
At this point, it does not matter why it happened but we need an urgent solution to solve this. 
So I see no reason to not have this feature on next.config.js.

 (and yes I do understand the points mentioned by @SukkaW  but we also need to understand that one size fits all solution does not fit all cases at all the time and usually we need an effective easy solution for real world projects where investigating cause and effect is too costly for any business).

(ps: if anyone have a quick solution regarding my issue, please suggest!)


![image](https://user-images.githubusercontent.com/9002363/195336308-88e25adc-3693-43a5-92e6-385e88ed8208.png)
